### PR TITLE
Workaround for cuFFT bug

### LINKED ISSUE #63152: torch.fft.rfft bug
## 🐛 Bug

When performing zero padding followed by rfft on cuda, incorrect results can happen on the second operation when padded length is even for longer sequences.  This behavior is observed on cuda only, not cpu.

## To Reproduce

```
import torch
import torch.fft as fft
import numpy as np
from scipy.fft import rfft


if __name__ == '__main__':
    dtype = torch.float64
    for device in ['cuda', 'cpu']:
        print('\ndevice={}'.format(device))
        Ly = 1600
        y = torch.randn(1, 1, Ly, dtype=dtype, device=device)
        for Lx in [160, 161, 1600, 1601, 16000, 16001]:
            x = torch.randn(1, 1, Lx, dtype=dtype, device=device)
            x = torch.nn.functional.pad(x, [0, Ly-1])
            yy = torch.nn.functional.pad(y, [0, Lx-1])
            X = fft.rfft(x)
            Y = fft.rfft(yy)
            X_s = rfft(x[0, 0, :].cpu().numpy())
            Y_s = rfft(yy[0, 0, :].cpu().numpy())
            dif_x = np.max(np.abs(X_s - X[0, 0, :].cpu().numpy()))
            dif_y = np.max(np.abs(Y_s - Y[0, 0, :].cpu().numpy()))
            print('Lx={}: dif_x={}, dif_y={}'.format(Lx, dif_x, dif_y))

```
Output:
```
device=cuda
Lx=160: dif_x=6.080392413311918e-12, dif_y=1.9477723502559983e-11
Lx=161: dif_x=1.4273168140722167e-14, dif_y=3.972054645195637e-14
Lx=1600: dif_x=8.206403321718826e-14, dif_y=113.69819800575286
Lx=1601: dif_x=5.0242958677880804e-14, dif_y=4.832213032818602e-14
Lx=16000: dif_x=3.0511999048007806e-13, dif_y=121.9687795227097
Lx=16001: dif_x=2.3198975494145925e-13, dif_y=5.54951617168815e-14

device=cpu
Lx=160: dif_x=2.416106516409301e-14, dif_y=7.105427357601002e-14
Lx=161: dif_x=1.1916163935586911e-14, dif_y=4.266955553143359e-14
Lx=1600: dif_x=7.105982447430966e-14, dif_y=9.500188217416325e-14
Lx=1601: dif_x=4.4938668397781776e-14, dif_y=4.278033777464534e-14
Lx=16000: dif_x=2.5579538487363607e-13, dif_y=7.8482010346102e-14
Lx=16001: dif_x=2.744571917763205e-13, dif_y=7.791709393054201e-14

```

## Expected behavior

dif_y for Lx=1601 and 16001 for cuda run should be around 1e-14.

## Environment

Collecting environment information...
PyTorch version: 1.9.0+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.10

Python version: 3.7.6 (default, Jan  8 2020, 19:59:22)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-4.19.56-1.el7.x86_64-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB

Nvidia driver version: 418.40.04
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] pytorch-ignite==0.2.0
[pip3] pytorch-lightning==1.1.1
[pip3] torch==1.9.0+cu111
[pip3] torchaudio==0.9.0
[pip3] torchinfo==1.5.3
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.10.0+cu111
[conda] numpy                     1.19.5                   pypi_0    pypi
[conda] pytorch-ignite            0.2.0                    pypi_0    pypi
[conda] pytorch-lightning         1.1.1                    pypi_0    pypi
[conda] torch                     1.9.0+cu111              pypi_0    pypi
[conda] torchaudio                0.9.0                    pypi_0    pypi
[conda] torchinfo                 1.5.3                    pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.10.0+cu111             pypi_0    pypi

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @ngimel @mruberry @peterbell10 @walterddr

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63328
* __->__ #63327

Fixes #63152

Differential Revision: [D30343558](https://our.internmc.facebook.com/intern/diff/D30343558)

### CODE REVIEW

File: aten/src/ATen/native/cuda/SpectralOps.cpp [hunk 1]
```diff
@@ -293,7 +309,18 @@ static const Tensor& _exec_fft(Tensor& out, const Tensor& self, IntArrayRef out_
   c10::optional<CuFFTConfig> uncached_plan;
   const CuFFTConfig * config = nullptr;
 
-  if (plan_cache.max_size() > 0) {
+  // Workaround for gh-63152, gh-58724
+  // Bluestein plans in CUDA 11.1 (cufft 10.3) cannot be re-used
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/SpectralOps.cpp:313]: cc @ptrblck (see linked issue for analysis)

File: aten/src/ATen/native/cuda/SpectralOps.cpp [hunk 2]
```diff
@@ -293,7 +309,18 @@ static const Tensor& _exec_fft(Tensor& out, const Tensor& self, IntArrayRef out_
...
+  bool use_caching = true;
+  if (10300 <= CUFFT_VERSION && CUFFT_VERSION < 10400) {
+    // Only cache plans for transforms with small prime factors
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/SpectralOps.cpp:316]: What about linking to https://docs.nvidia.com/cuda/cufft/index.html#accuracy-and-performance here? 

Also, for non-expert readability, we probably want to say that Bluestein plans are only used when that condition is met and that other plans are reusable, hence the check

File: aten/src/ATen/native/cuda/SpectralOps.cpp [hunk 3]
```diff
@@ -234,6 +234,22 @@ void cufft_clear_plan_cache_impl(int64_t device_index) {
 namespace {
 constexpr int64_t cufft_max_ndim = 3;
 
+// "Large" here means a prime factor not special-cased by cuFFT
+bool has_large_prime_factor(int64_t n) {
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/SpectralOps.cpp:238]: Elaborate in this comment on the intended use of this function and what constitutes a "large" prime factor (and why that's "large" in this context)

File: aten/src/ATen/native/cuda/SpectralOps.cpp [hunk 4]
```diff
@@ -234,6 +234,22 @@ void cufft_clear_plan_cache_impl(int64_t device_index) {
...
+  constexpr int64_t first_large_prime = 11;
+  const std::array<int64_t, 4> prime_radices{{2, 3, 5, 7}};
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/SpectralOps.cpp:240]: We compile with C++14 so I don't think the double braces are needed?

File: aten/src/ATen/native/cuda/SpectralOps.cpp [hunk 5]
```diff
@@ -234,6 +234,22 @@ void cufft_clear_plan_cache_impl(int64_t device_index) {
...
+  for (auto prime : prime_radices) {
+    if (n < first_large_prime) {
+        return false;
+    }
+
+    while (n % prime == 0) {
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/SpectralOps.cpp:246]: Nit (doesn't have to be changed): In general compilers do a good job of optimizing modulus operations followed by division but this might actually be a case for std::div

File: aten/src/ATen/native/cuda/SpectralOps.cpp [hunk 6]
```diff
@@ -293,7 +309,20 @@ static const Tensor& _exec_fft(Tensor& out, const Tensor& self, IntArrayRef out_
   c10::optional<CuFFTConfig> uncached_plan;
   const CuFFTConfig * config = nullptr;
 
-  if (plan_cache.max_size() > 0) {
+  // Workaround for gh-63152, gh-58724
+  // Bluestein plans in CUDA 11.1 (cufft 10.3) cannot be re-used
+  bool use_caching = true;
+#ifdef CUFFT_VERSION
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/SpectralOps.cpp:315]: Is it possible for types like `CuFFTParams` to be defined but CUFFT_VERSION be undefined?

**Comment by mruberry** [comment]
This is really cool, @peterbell10, nice fix. I made a few small inline suggestions to extend the comments

File: aten/src/ATen/native/cuda/SpectralOps.cpp (hunk 4 above)
Comment (peterbell10) [reply aten/src/ATen/native/cuda/SpectralOps.cpp:240]: I think all compilers accept single braces here, but technically it should be two since it's an aggregate initizalizer with an array member.

File: aten/src/ATen/native/cuda/SpectralOps.cpp (hunk 5 above)
Comment (peterbell10) [reply aten/src/ATen/native/cuda/SpectralOps.cpp:246]: `std::div` is often slower than just letting the compiler optimize it, since `std::div` is behind a library function call.

File: aten/src/ATen/native/cuda/SpectralOps.cpp (hunk 6 above)
Comment (peterbell10) [reply aten/src/ATen/native/cuda/SpectralOps.cpp:315]: ROCm with hipFFT.

File: aten/src/ATen/native/cuda/SpectralOps.cpp (hunk 2 above)
Comment (peterbell10) [reply aten/src/ATen/native/cuda/SpectralOps.cpp:320]: I've added the link on `has_large_prime_factor` and added a comment about small prime factors being okay.

File: aten/src/ATen/native/cuda/SpectralOps.cpp (hunk 4 above)
Comment (mruberry) [reply aten/src/ATen/native/cuda/SpectralOps.cpp:241]: OK

**Review by mruberry — APPROVED** [review]
Cool! Nice surgical fix, @peterbell10!

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30343558).
